File: Python/String/0003_longest_substring_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Base Cases
        if (s == ''):
            return 0
        if (len(s) == 1):
            return 1
        
        dicSub = {}
        left = 0
        largest = 0
        
        for i in range(len(s)):
            if (s[i] in dicSub):
                largest = max(largest, i - left)
                left = max(left, dicSub[s[i]] + 1)
            dicSub[s[i]] = i
            
        return max(largest, len(s) - left)
    
        '''
        # Using for loop
        
        # Base Cases
        if (s == ''):
            return 0
        if (len(s) == 1):
            return 1
        
        count = 1
        largest = 1
        before = 0
        for i in range(1, len(s)):
            if (s[i] in s[before:i]):
                largest = max(largest, count)
                count = 1
                before = i - 1
            else:
                count = i - before + 1
        
        largest = max(largest, count)
        
        return largest
        '''

File: Swift/String/0003_longest_substring_without_repeating_characters.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        // Base Cases
        if (s == "") {
            return 0
        }
        if (s.count == 1) {
            return 1
        }
        
        //var dicSub = ["star": 0]
        var dicSub: [Character: Int] = [:]
        var left = 0
        var largest = 0
        
        for (i, char) in s.enumerated() {
            if (dicSub[char] != nil) {
                largest = max(largest, i - left)
                left = max(left, dicSub[char]! + 1)
            }
            dicSub.updateValue(i, forKey: char)
        }
            
        return max(largest, s.count - left)
    }
}
